library/core/src/sync/atomic.rs

```diff
@@ -820,6 +842,30 @@ impl<T> AtomicPtr<T> {
         unsafe { &mut *self.p.get() }
     }
 
+    /// Get atomic access to a pointer.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(atomic_from_mut)]
+    /// use std::sync::atomic::{AtomicPtr, Ordering};
+    ///
+    /// let mut some_ptr = &mut 123 as *mut i32;
+    /// let a = AtomicPtr::from_mut(&mut some_ptr);
+    /// a.store(&mut 456, Ordering::Relaxed);
+    /// assert_eq!(unsafe { *some_ptr }, 456);
+    /// ```
+    #[inline]
+    #[unstable(feature = "atomic_from_mut", issue = "76314")]
+    pub fn from_mut(v: &mut *mut T) -> &Self {
+        let [] = [(); align_of::<AtomicPtr<()>>() - align_of::<*mut ()>()];
```

Comment:
This will assert that their alignment is *equal* right? But shouldn't we rather test that `Self` alignment is no stricter than that of the pointer? You can use the [`static_assert` macro](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_data_structures/macros.rs.html#4-11) to write any kind of boolean test.

(This is what causes the libc CI failure, but I am not sure if the inequality there is in the right direction.)

Reply:
Btw, if you know of any other place that uses the such an array length hack to encode a static assertion, please let me know -- we should add a FIXME at least.

Reply:
`get_mut` already assumes that the alignment of the non-atomic type is ≤ the alignment of the atomic variant. `from_mut` technically only needs ≥, but things would be quite broken if it was > instead of just =.

Reply:
> You can use the [`static_assert` macro](https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_data_structures/macros.rs.html#4-11) to write any kind of boolean test.

Can I depend on that in `core`? I don't think `core` has any dependencies right now.

Reply:
Oh right, you cannot directly use it.

But personally I feel like it'd be worth copying it to libcore and then using it there as an internal macro. That said, it seems you will have to rework this entire thing anyway so let's see where that goes.